Add optimizers from nevergrad

gauravmanmode opened the pull request:
## PR Description
This PR adds support for the following optimizers from the [nevergrad](https://facebookresearch.github.io/nevergrad/index.html) optimization library.

- [x] PSO
- [x] CMAES
- [x] ONEPLUSONE
- [x] RANDOMSEARCH
- [x] SAMPLINGSEARCH
- [x] DE
- [x] BO
- [x] EDA
- [x] TBPSA
- [x] EMNA
- [x] NGOPT OPTIMIZERS
- [x] META OPTIMIZERS 

Two optimizers from nevergrad are not wrapped namely SPSA and AXP as they are either slow or imprecise.

## Features:

- [x] Parallelization
- [ ] Support for nonlinear constraints

## Helper functions 
`_nevergrad_internal`:
 Handle the optimization loop, return InternalOptimizeResult .

`x _process_nonlinear_constraints`:
Flatten vector constraint into a list of scalar constraints for use with Nevergrad.

`x _get_constraint_evaluations`:
Return a list of constraint evaluations at x

`x _batch_constraint_evaluations`:
  Batch version of _get_constraint_evaluations

## Test suite:
```
x test_process_nonlinear_constraints
x test_get_constraint_evaluations
x test_batch_constraint_evaluations
test_meta_optimizers_are_valid
test_ngopt_optimizers_are_valid
```
Note:
Nonlinear constraints on hold until improved handling.

Changes to `optimize.py`:
Currently None bounds are transformed to arrays of np.inf. Handle this case if optimizer does not support infinite bounds.

Added test test_infinite_and_incomplete_bounds.py:
`test_no_bounds_with_nevergrad`
This test should pass when no bounds are provided to nevergrad optimizers.

janosg commented:
Hi @gauravmanmode, thanks for the PR. 

I definitely like the idea of your `nevergrad_internal` function. We currently have several independent nevergrad PRs open and a function like this is good to avoid code duplication. 

Regarding the Executor: There was an [argument brought forward by @r3kste](https://github.com/optimagic-dev/optimagic/issues/560#issuecomment-2798944370) that suggests it would be better to use the low-level ask-and-tell interface if we want to support parallelism. While I still think the solution with the custom Executor can be made to work, I think that the ask-and-tell interface is simpler and more readable for this.

janosg commented:
Currently your tests fail because nevergrad is not compatible with numpy 2.0 and higher. You can pin numpy in the environment file for now.

janosg commented:
Or better: Install nevergrad via pip instead of conda. The conda version is outdated. Then you don't need to pin any numpy versions.

gauravmanmode commented:
Hi, @janosg ,
Installing nevergrad with pip solved the failing tests.

Here is the list of parameter names I have referred to 
### nevergrad_cmaes
| **Old Name**        | **Proposed Name**                        | **from optimizer in optimagic** |
|---------------------|------------------------------------------|----------------------------------|
| `tolx`              | `xtol`                                   | `scipy`                  |
| `tolfun`            | `ftol`                                   | `scipy`                        |
| `budget`       | `stopping_maxfun`            |   `scipy`               |
| `CMA_rankmu`        | `learning_rate_rank_mu_update`          | `pygmo_cmaes`                  |
| `CMA_rankone`       | `learning_rate_rank_one_update`         | `pygmo_cmaes`              |
| `popsize`           | `population_size`                       | `pygmo_cmaes`                  |
| `fcmaes`            | `use_fast_implementation`               | `needs review`                  |
| `diagonal`          | `diagonal`                              |  `needs review`               |
| `elitist`           | `elitist`                               |  `needs review`               |
| `scale`           | `scale`                               |  `needs review`               |
| `num_workers`           | `n_cores`                               |  `needs review`               |
| `high_speed`        | `high_speed`                            |  `needs review`            |

what kind of tests should i have for the internal helper function ?
Should I have tests for ftol, stopping_maxfun?

gauravmanmode commented:
Hi @janosg,
I am thinking of refactoring the code for the already added `nevergrad_pso` optimizer and `nevergrad_cmaes` in this pr.
Does this sound good?
Also, I would like your thoughts on this.
1. currently I am passing the optimizer object to the helper function `_nevergrad_internal`.
![image](https://github.com/user-attachments/assets/5824b34d-8cb6-4c46-aef1-f32902bd285a)
2. Another approach is to pass the optimizer name as a string as in `pygmo`
![image](https://github.com/user-attachments/assets/e29d15ef-b74f-43c9-84cd-6fa977b5ac00)
![image](https://github.com/user-attachments/assets/b89b43f1-fe89-4265-b5df-c090ae39f760)
What would be a better choice?

janosg commented:
Hi @gauravmanmode, yes please go ahead and refactor the code for pso as well. 

I would stick to approach one, i.e. passing the configured optimizer object to the internal function. It is more in line with the design philosophy shown [here](https://optimagic.readthedocs.io/en/latest/how_to/how_to_specify_algorithm_and_algo_options.html).

janosg commented:
> Hi, @janosg , Installing nevergrad with pip solved the failing tests.
> 
> Here is the list of parameter names I have referred to
> 
> ### nevergrad_cmaes
> **Old Name**	**Proposed Name**	**from optimizer in optimagic**
> `tolx`	`xtol`	`scipy`
> `tolfun`	`ftol`	`scipy`
> `budget`	`stopping_maxfun`	`scipy`
> `CMA_rankmu`	`learning_rate_rank_mu_update`	`pygmo_cmaes`
> `CMA_rankone`	`learning_rate_rank_one_update`	`pygmo_cmaes`
> `popsize`	`population_size`	`pygmo_cmaes`
> `fcmaes`	`use_fast_implementation`	`needs review`
> `diagonal`	`diagonal`	`needs review`
> `elitist`	`elitist`	`needs review`
> `seed`	`seed`	
> `scale`	`scale`	`needs review`
> `num_workers`	`n_cores`	`optimagic`
> `high_speed`	`high_speed`	`needs review`
> what kind of tests should i have for the internal helper function ? Should I have tests for ftol, stopping_maxfun? Also, in nevergrad, recommendation.loss returns None for some optimizers like CMA. Is this a nevergrad issue or am i missing something. ![image](https://private-user-images.githubusercontent.com/44936435/440350402-de852b84-4b38-47b0-bc24-a7c347b44e25.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY4NjEzMTYsIm5iZiI6MTc0Njg2MTAxNiwicGF0aCI6Ii80NDkzNjQzNS80NDAzNTA0MDItZGU4NTJiODQtNGIzOC00N2IwLWJjMjQtYTdjMzQ3YjQ0ZTI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTEwVDA3MTAxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllMzBhNmZiN2FjZjU5OTVlODU4ZGM1YjY2ZDY5MWQwMGVkOTM0ZDA4MjFlNzY4ZjllMjc1ZDk5NDljODc0NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Z8RuGiSD4i4JK7xoJHq1jAM9PW66rkcFT4SA1LuBa08) For reference, I have attached a notebook I used while exploring [here](https://github.com/gauravmanmode/sharing/blob/main/test_nevergrad.ipynb)

About the names:

- `xtol` and `ftol` are convergence criteria, so the name would be `convergence_xtol`. Ideally you would also find out if this is an absolute or relative tolerance and then add the corresponding abbreviation (e.g. `convergence_xtol_rel`). You can find examples of the naming scheme [here](https://github.com/optimagic-dev/optimagic/blob/main/src/optimagic/optimization/algo_options.py)
- The otrher names are god

I would mainly add a name for stopping_maxfun. Other convergence criteria are super hard to test. 

If you cannot get a loss out of nevergrad for some optimizers you can evaluate problem.fun at the solution for now and create an issue with a minimal example at nevergrad to get feedback. I wouldn't frame it as a bug report (unless you are absolutely sure) but rather frame it as a question whether you are using the library correctly.

gauravmanmode commented:
##  Update
Refactored the code and improved type annotations.
Unable to get optimal loss in case of CMAES optimizer. Here is a related [issue](https://github.com/facebookresearch/nevergrad/issues/703). I have also created a issue at nevergrad [here.](https://github.com/facebookresearch/nevergrad/issues/1697) .
Keeping in line with the styleguide, I have added the algo info in the docstrings itself.
These are the additional options:
| **Old Name**          | **Proposed Name**                    |
|-----------------------|--------------------------------------|
| scale                 | scale                               |
| elitist               | elitist                             |
| popsize               | population_size                     |
| diagonal              | diagonal                            |
| high_speed            | high_speed                          |
| fcmaes                | fast_cmaes                          |
| random_init           | random_init                         |
| AdaptSigma            | step_size_adaptive                  |
| CMA_active            | negative_update                     |
| CMA_cmean             | learning_rate_mean_update           |
| CMA_const_trace       | normalize_cov_trace                 |
| CMA_diagonal          | diag_covariance_iters               |
| CMA_diagonal_decoding | learning_rate_diagonal_update       |
| CMA_mirrormethod      | mirror_sampling_strategy            |
| CMA_mu                | num_parents                         |
| CMA_on                | learning_rate_cov_mat_update        |
| CMA_rankmu            | learning_rate_rank_mu_update        |
| CMA_rankone           | learning_rate_rank_one_update       |
| CMA_dampsvec_fade     | step_size_damping_rate              |
| CSA_dampfac           | step_size_damping_factor            |
| CSA_squared           | step_size_update_squared            |
| CSA_invariant_path    | CSA_invariant_path                  |
| eval_final_mean       | eval_final_mean                     |
| maxfevals             | stopping_maxfun                     |
| maxiter               | stopping_maxiter                    |
| timeout               | stopping_timeout                    |
| tolconditioncov       | stopping_cov_mat_cond               |
| tolfun                | convergence_ftol_abs                |
| tolfunrel             | convergence_ftol_rel                |
| tolstagnation         | convergence_iter_noimprove          |
| tolx                  | convergence_xtol_abs                |

janosg commented:
Hi @gauravmanmode, thanks for the update. For all names that start with CSA_ or CMA I would stick to the original. Your names are a bit more descriptive but I would still need a longer description to understand what they do and what values they can take. So we can stick to the originals to make it easier to switch from nevergrad and put the rest into the documentation.

janosg commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -77,34 +107,264 @@ def _solve_internal_problem(
+    convergence_ftol_abs: float = 1e-11
+    convergence_ftol_rel: float = 0.0
+    convergence_xtol_abs: float = 1e-11
+    convergence_iter_noimprove: int | None = None
```
PositiveInt?

gauravmanmode commented on .pre-commit-config.yaml:
```diff
@@ -29,6 +29,7 @@ repos:
         additional_dependencies:
           - hatchling
           - ruff
+          - iminuit
```
Sorry for that. I had some issue locally so I had it temporarily. Will remove from here.

gauravmanmode commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -77,34 +107,264 @@ def _solve_internal_problem(
+    random_init: bool = False
+    n_cores: PositiveInt = 1
+    step_size_adaptive: bool | str = True
+    step_size_damping_factor: float = 1.0
+    step_size_damping_rate: float = 0.1
+    step_size_update_squared: bool = False
+    learning_rate_cov_mat_update: float = 1.0
+    learning_rate_rank_one_update: float = 1.0
+    learning_rate_rank_mu_update: float = 1.0
+    learning_rate_mean_update: float = 1.0
```
I had some changes made locally. I will commit them soon and improve on the type hints.

gauravmanmode commented on src/optimagic/optimizers/nevergrad_optimizers.py:
```diff
@@ -77,34 +107,264 @@ def _solve_internal_problem(
+        learning_rate_rank_one_update: Multiplier for the rank-one update learning rate
+            of the covariance matrix.
+        learning_rate_rank_mu_update: Multiplier for the rank-mu update learning rate
+            of the covariance matrix.
+        learning_rate_mean_update: Learning rate for the mean update.
+        learning_rate_diagonal_update: Learning rate for the diagonal update.
+        num_parents: Number of parents(μ) for recombination.
+        negative_update: Whether to use negative updates for the covariance matrix.
+        mirror_sampling_strategy: Strategy for mirror sampling. Possible values are:
+            0: Unconditional mirroring
+            1: Selective mirroring
+            2: Selective mirroring with delay (default)
+        normalize_cov_trace: How to normalize the trace of the covariance matrix.
+            Valid values are:
+            - False: No normalization
+            - True: Normalize to 1
+            - "arithm": Arithmetic mean normalization
+            - "geom": Geometric mean normalization
+            - "aeig": Arithmetic mean of eigenvalues
+            - "geig": Geometric mean of eigenvalues
+        diag_covariance_iters: Number of iterations to use diagonal covariance matrix
+            before switching to full matrix. If False, always use full matrix.
+        stopping_maxfun: Maximum number of function evaluations before termination.
+        stopping_maxiter: Maximum number of iterations before termination.
+        stopping_timeout: Maximum time in seconds before termination.
+        stopping_cov_mat_cond: Maximum condition number of the covariance matrix before
+            termination.
+        convergence_ftol_abs: Absolute tolerance on function value changes for
+            convergence.
+        convergence_ftol_rel: Relative tolerance on function value changes for
+            convergence.
+        convergence_xtol_abs: Absolute tolerance on parameter changes for convergence.
+        convergence_iter_noimprove: Number of iterations without improvement before
+            termination.
+        invariant_path: Whether evolution path (pc) should be invariant to
+            transformations.
+        eval_final_mean: Whether to evaluate the final mean solution.
+        seed: Seed used by the internal random number generator for reproducibility.
+
+    """
```
Sure, it would be redundant to have it at both places, but in the styleguide it was mentioned to be preferred adding in the docstrings, so I had it both places.

gauravmanmode commented:
## Update
I have updated the PR Description .
These are the parameter names for the additonal optimizers added.

<details>
<summary><strong>RandomSearch</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `stupid`            | `baseline`          |
| `middle_point`      | `init_zero`         |
| `opposition_mode`   | `mirror_sampling`   |
| `sampler`           | `sampling_method`   |
| `scale`             | `scale`             |
| `recommendation_rule`| `recommendation_rule`|
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>SamplingSearch</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `sampler`           | `sampling_method`   |
| `scrambled`         | `scrambled`         |
| `middle_point`      | `init_zero`         |
| `cauchy`            | `cauchy`            |
| `scale`             | `scale`             |
| `rescaled`          | `rescaled`          |
| `recommendation_rule`| `recommendation_rule`|
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>TBPSA</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `naive`             | `naive`             |
| `initial_popsize`   | `initial_popsize`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>EMNA</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `isotropic`         | `isotropic`         |
| `naive`             | `naive`             |
| `population_size_adaptation`| `population_size_adaptation`|
| `initial_popsize`   | `initial_popsize`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>BayesOptim</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `init_budget`       | `init_budget`       |
| `pca`               | `pca`               |
| `n_components`      | `n_components`      |
| `prop_doe_factor`   | `prop_doe_factor`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>Differential Evolution (DE)</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `initialization`    | `initialization`    |
| `scale`             | `scale`             |
| `recommendation`    | `recommendation`    |
| `crossover`         | `crossover`         |
| `F1`                | `F1`                |
| `F2`                | `F2`                |
| `population_size`   | `population_size`   |
| `high_speed`        | `high_speed`        |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

gauravmanmode commented:
## Update
I have updated the PR Description .
These are the parameter names for the additonal optimizers added.

I have made a few changes to non_linear constraints and test_history_collection to work with the added optimizers, but they can be reverted back.
nonlinear_constraints https://github.com/optimagic-dev/optimagic/issues/606
As there is no stopping_maxiter option with optimizers in nevergrad, stopping_maxfun would be a good way to limit the numbe of entries in test_history_collection 

<details>
<summary><strong>RandomSearch</strong></summary>
nevergrad_randomsearch
| Old Name         | New Name         |
|------------------|------------------|
| `stupid`            | `baseline`          |
| `middle_point`      | `init_zero`         |
| `opposition_mode`   | `mirror_sampling`   |
| `sampler`           | `sampling_method`   |
| `scale`             | `scale`             |
| `recommendation_rule`| `recommendation_rule`|
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>SamplingSearch</strong></summary>
nevergrad_samplingsearch
| Old Name         | New Name         |
|------------------|------------------|
| `sampler`           | `sampling_method`   |
| `scrambled`         | `scrambled`         |
| `middle_point`      | `init_zero`         |
| `cauchy`            | `cauchy`            |
| `scale`             | `scale`             |
| `rescaled`          | `rescaled`          |
| `recommendation_rule`| `recommendation_rule`|
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>TBPSA</strong></summary>
nevergrad_tbpsa
| Old Name         | New Name         |
|------------------|------------------|
| `naive`             | `naive`             |
| `initial_popsize`   | `initial_popsize`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>EMNA</strong></summary>
nevergrad_emna
| Old Name         | New Name         |
|------------------|------------------|
| `isotropic`         | `isotropic`         |
| `naive`             | `naive`             |
| `population_size_adaptation`| `population_size_adaptation`|
| `initial_popsize`   | `initial_popsize`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>BayesOptim</strong></summary>
nevergrad_bayes_optim
| Old Name         | New Name         |
|------------------|------------------|
| `init_budget`       | `init_budget`       |
| `pca`               | `pca`               |
| `n_components`      | `n_components`      |
| `prop_doe_factor`   | `prop_doe_factor`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
<summary><strong>Differential Evolution (DE)</strong></summary>
`nevergrad_de`

| Old Name         | New Name         |
|------------------|------------------|
| `initialization`    | `initialization`    |
| `scale`             | `scale`             |
| `recommendation`    | `recommendation`    |
| `crossover`         | `crossover`         |
| `F1`                | `F1`                |
| `F2`                | `F2`                |
| `population_size`   | `population_size`   |
| `high_speed`        | `high_speed`        |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

gauravmanmode commented:
## Update
I have updated the PR Description .
These are the parameter names for the additonal optimizers added.

I have made a few changes to non_linear constraints and test_history_collection to work with the added optimizers, but they can be reverted back.
nonlinear_constraints https://github.com/optimagic-dev/optimagic/issues/606
As there is no stopping_maxiter option with optimizers in nevergrad, stopping_maxfun would be a good way to limit the numbe of entries in test_history_collection 

<details>
nevergrad_randomsearch
<summary><strong>RandomSearch</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `stupid`            | `baseline`          |
| `middle_point`      | `init_zero`         |
| `opposition_mode`   | `mirror_sampling`   |
| `sampler`           | `sampling_method`   |
| `scale`             | `scale`             |
| `recommendation_rule`| `recommendation_rule`|
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
nevergrad_samplingsearch
<summary><strong>SamplingSearch</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `sampler`           | `sampling_method`   |
| `scrambled`         | `scrambled`         |
| `middle_point`      | `init_zero`         |
| `cauchy`            | `cauchy`            |
| `scale`             | `scale`             |
| `rescaled`          | `rescaled`          |
| `recommendation_rule`| `recommendation_rule`|
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
nevergrad_tbpsa
<summary><strong>TBPSA</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `naive`             | `naive`             |
| `initial_popsize`   | `initial_popsize`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
nevergrad_emna
<summary><strong>EMNA</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `isotropic`         | `isotropic`         |
| `naive`             | `naive`             |
| `population_size_adaptation`| `population_size_adaptation`|
| `initial_popsize`   | `initial_popsize`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
nevergrad_bayes_optim
<summary><strong>BayesOptim</strong></summary>

| Old Name         | New Name         |
|------------------|------------------|
| `init_budget`       | `init_budget`       |
| `pca`               | `pca`               |
| `n_components`      | `n_components`      |
| `prop_doe_factor`   | `prop_doe_factor`   |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

<details>
`nevergrad_de`
<summary><strong>Differential Evolution (DE)</strong></summary>


| Old Name         | New Name         |
|------------------|------------------|
| `initialization`    | `initialization`    |
| `scale`             | `scale`             |
| `recommendation`    | `recommendation`    |
| `crossover`         | `crossover`         |
| `F1`                | `F1`                |
| `F2`                | `F2`                |
| `population_size`   | `population_size`   |
| `high_speed`        | `high_speed`        |
| `stopping_maxfun`   | `stopping_maxfun`   |
| `n_cores`           | `n_cores`           |
| `seed`              | `seed`              |
| `sigma`             | `sigma`             |

</details>

---

gauravmanmode commented on tests/optimagic/optimization/test_many_algorithms.py:
```diff
@@ -162,6 +162,8 @@ def test_deterministic_global_algorithm_on_sum_of_squares(algorithm):
 @pytest.mark.skipif(sys.platform == "win32", reason=skip_msg)
 @pytest.mark.parametrize("algorithm", GLOBAL_STOCHASTIC_ALGORITHMS)
 def test_stochastic_global_algorithm_on_sum_of_squares(algorithm):
+    if algorithm in ("nevergrad_axp"):
```
It is slow for stopping_maxfun = 1000 takes ~3 mins. stopping_maxfun = 100 might work.

janosg commented:
Can you quickly explain why you removed SPSA?

gauravmanmode commented:
SPSA was not accurate and was failing the tests.

janosg commented:
This is something we always need to discuss before we decide to drop an algorithm. Often it is possible to tune the parameters to make algorithms more precise; In extreme cases we can also relax the required precision for algorithms before we drop them. 

I merged main into your branch. Now tests are failing due to the changes in #610 but this will be a quick fix.

gauravmanmode commented:
Sorry I missed a discussion on this. 
But the implentation of SPSA in nevergrad was a WIP (many todos listed) and no tuning parameters exposed, so I decided to skip it

janosg commented on tests/optimagic/optimizers/test_nevergrad.py:
```diff
@@ -0,0 +1,133 @@
+# test stochastic_global_algorithm_on_sum_of_squares
+@pytest.mark.slow
+@pytest.mark.parametrize("algorithm", NEVERGRAD_META)
+def test_meta_optimizers_with_stochastic_global_algorithm_on_sum_of_squares(algorithm):
+    res = minimize(
+        fun=sos,
+        params=np.array([0.35, 0.35]),
+        bounds=Bounds(lower=np.array([0.2, -0.5]), upper=np.array([1, 0.5])),
+        algorithm=algorithms.NevergradMeta(algorithm),
+        collect_history=False,
+        skip_checks=True,
+        algo_options={"seed": 12345},
+    )
+    assert res.success in [True, None]
+    aaae(res.params, np.array([0.2, 0]), decimal=1)
```
Why do we need this test, especially if it is slow. It looks very similar to the tests in `test_many_algortihms`? Which case does it test that is not tested there?

janosg replied:
I remember now. I think you have told me before. Sorry.

janosg commented on tests/optimagic/optimizers/test_nevergrad.py:
```diff
@@ -0,0 +1,133 @@
+
+@pytest.mark.slow
+@pytest.mark.parametrize("algorithm", NEVERGRAD_NGOPT)
+def test_ngopt_optimizers_with_stochastic_global_algorithm_on_sum_of_squares(algorithm):
```
Same. Why do we need this test?

gauravmanmode replied:
same

janosg reviewed (approved):
Hi @gauravmanmode, thanks for the great PR! This looks great! I have two small comments about tests you wrote but I already approve the PR.

timmens commented:
Hey @gauravmanmode. I built the environment locally. You are right that nevergrad=1.0.3 is installed, which [expects bayesian-optimization>=1.2.0](https://github.com/facebookresearch/nevergrad/blob/c23f20406d9223afbe253f9807ca7e8ba993a1ac/requirements/main.txt). I suspect that nevergrad can't work with the breaking change of bayesian-optimization (from 1.x.y to 2.x.y). If you try to import nevergrad in a local environment you will see that this fails because of bayesian-optimization. This is why the tests fail.

I think there are two routes we can take:

1. We support and use bayesian-optimization < 2 in our own development environment. In this case we would need to make sure that everything works with version<2 and version>=2 (I am not sure if this has been checked; maybe @spline2hg can say something about this?).
2. We remove nevergrad from the dev environment and test it in a separate environment where we can install bayesian-optimization=1.4. This is a bit annoying right now, but will be a very easy solution once we switch to a modern package manager.

janosg commented:
@timmens lets go for option 2. Could you do those changes since you have done similar changes many times before?

@gauravmanmode could you open an issue at nevergrad to see if there are fundamental reasons why the version needs to be pinned so strictly? Otherwise we could offer to help them make nevergrad compatible with newer versions of bayesian optimization.

gauravmanmode commented:
Sure, I’ll open an issue on Nevergrad. As @timmens told, the reason the version is pinned strictly is because nevergrad uses some deprecated functions and not updated since.

timmens commented:
I'll try to explain what you need to do for "Route 2"  here @gauravmanmode (seems like the easiest solution as I don't have push access to your branch I believe).

1. You remove nevergrad from our development environments. I'd suggest commenting it out and leaving a comment saying that it is currently incompatible with bayesian-optimization.
2. You edit `.tools/update_envs.py`:
    1. You create a new section for "nevergrad"
    2. Here you remove the bayesian-optimization (much like we remove the numpy or pandas lines for the test environments for numpy < 2)
    3. You insert new lines for "nevergrad" and "bayesian-optimization==1.4" in the index for the pip versions.
    4. You add this new environment to the loop "write environments" under the name "nevergrad".
 3. In `main.yml` you add a section `run-nevergrad-tests`. Make sure that the pytest call only calls tests that are related to nevergrad.
 
 And then you should be done!

spline2hg commented:
I see the issue is resolved. Just to clarify the currently implemented Bayesian optimizer is not compatible with `bayesian-optimization==1.4`. Specifically, the acquisition function we rely on wasn’t properly introduced until version 2.0.0, so the current approach(option 2) used seems appropriate.thanks!

gauravmanmode commented:
A related issue on nevergrad https://github.com/facebookresearch/nevergrad/issues/1701.
